Adjust use of HTTPS vs. "https", etc. (#341)
MikeBishop/dns-alt-svc: draft-ietf-dnsop-svcb-https.md

This document specifies the "SVCB" and "HTTPS" DNS resource record (RR)
types to facilitate the lookup of information needed to make connections

> to network services, such as for "https" origins.  SVCB records

Reviewer: ```suggestion
to network services, such as for HTTP origins.  SVCB records
```

allow a service to be provided from multiple alternative endpoints,
each with associated parameters (such as transport protocol
configuration and keys for encrypting the TLS ClientHello).  They also
enable aliasing of apex domains, which is not possible with CNAME.

> The HTTPS RR is a variation of SVCB for "https" and "http" origins.

Reviewer: ```suggestion
The HTTPS RR is a variation of SVCB for use with HTTP {{!HTTP=I-D.ietf-httpbis-semantics}}.
```

public keys.

For example, when clients need to make a connection to fetch resources

> associated with an "https" URI, they currently resolve only A and/or AAAA

Reviewer: ```suggestion
associated with a URI with an "https" or "http" scheme, they currently resolve only A and/or AAAA
```

Author: This is an example, and the next part is specific to "https" origins, so I don't know if it makes sense to list both schemes here.

Author: The next sentence is
```
This is adequate for services that use
basic HTTP/TLS (fixed port, no QUIC, no {{!ECH=I-D.ietf-tls-esni}}), but
clients that learn additional information can ...
```
Clients don't use HTTP/TLS when the URI has an "http" scheme, so this becomes a non sequitur ... unless we're talking about RFC 8164, which is another layer of complexity that I would prefer not to raise.

This is just supposed to be an example, so there's no need to make it comprehensive.  How about: "For example, when accessing an "https" URI, clients currently resolve only ..."?

Author: OK, how about

> For example, HTTP clients currently resolve only A and/or AAAA records for the origin hostname, learning only its IP addresses.  If HTTP clients learn more information about the origin before connecting, they may be able to upgrade "http" URLs to "https", enable HTTP/3 or ECH, or switch to an operationally preferable endpoint.

Reviewer: Works for me.

Author: Done

own value-list.  Other automatically mandatory keys SHOULD NOT appear in the
list either.  (Including them wastes space and otherwise has no effect.)

> # Service Bindings for "https" {#https}

Author: My argument here is that a Service Binding mapping is specific to a scheme, not a protocol, and this section only defines a mapping for the "https" scheme.  Use of an "http" scheme with HTTPS or SVCB is actually forbidden!

Conversely, if you wanted to use some protocol that is not HTTP but somehow relies on the "https" scheme (maybe e.g. secure websockets?), this section would still apply.

> # Service Bindings for "https" {#https}

Reviewer: But you still have to do the lookup with "http"-schemed URIs, or you won't discover the HSTS-like directive.  So this does apply to all HTTP usage, just differently.

> # Service Bindings for "https" {#https}

Author: It does apply to "http", but it doesn't create a service binding for "http".  The implications for "http" usage are a special behavior of the HTTPS RR type, separate from the Service Binding functionality, which is only for "https".

How about "Using Service Bindings with HTTP", or just "The HTTPS RR Type"?

> # Service Bindings for "https" {#https}

Reviewer: "Using Service Bindings with HTTP" sounds good.

> # Service Bindings for "https" {#https}

Author: Done

The "automatically mandatory" keys ({{mandatory}}) are "port"
and "no-default-alpn".  (As described in {{mandatory}}, clients must
either implement these keys or ignore any RR in which they appear.)
Clients that restrict the destination port in "https" URIs
(e.g. using the "bad ports" list from {{FETCH}}) SHOULD apply the
same restriction to the "port" SvcParam.

The presence of an HTTPS RR for an origin also indicates
that clients should connect securely and use the "https" scheme, as
discussed in {{hsts}}.  This allows HTTPS RRs to apply to
pre-existing "http" scheme URLs, while ensuring that the client uses a

> secure and authenticated HTTPS connection.

Reviewer 2: ```suggestion
secure and authenticated "https" scheme connection.
```

If an HTTPS RR query for this "https" URL returns any AliasMode HTTPS RRs,
or any compatible ServiceMode HTTPS RRs (see {{mandatory}}), the client

> SHOULD act as if it has received an HTTP 307 "Temporary Redirect"
> to this "https" URL.  (Receipt of an incompatible ServiceMode RR does not

Reviewer: ```suggestion
SHOULD behave as if it has received an HTTP 307 (Temporary Redirect) status code
with this "https" URL in the "Location" field.  (Receipt of an incompatible ServiceMode RR does not
```

resources that are only available on the "http" origin), the operator MUST
NOT publish an HTTPS RR.

> When an "https" connection fails due to an error in the underlying secure

Reviewer: ```suggestion
When an HTTP connection fails due to an error in the underlying secure
```

Author: Given that HTTP connections don't generically have a secure transport, I find this version harder to understand.

## HTTPS RRType

This document defines a new DNS RR type, "HTTPS", whose value 65 has
been allocated by IANA from the "Resource Record (RR) TYPEs"
subregistry of the "Domain Name System (DNS) Parameters" registry:

* Type: HTTPS
* Value: 65

> * Meaning: Service Binding for "https"

Reviewer: ```suggestion
* Meaning: Service Binding for HTTP
```

Reviewer: As before, I prefer the last option since it affects all HTTP requests.

Author: Done
